winit/src/conversion.rs

```diff
@@ -156,6 +157,119 @@ pub fn visible(mode: Mode) -> bool {
+    modifiers.set(ModifiersState::CTRL, hotkey.modifiers.control());
+    modifiers.set(ModifiersState::SHIFT, hotkey.modifiers.shift());
+    modifiers.set(ModifiersState::ALT, hotkey.modifiers.alt());
+    modifiers.set(ModifiersState::LOGO, hotkey.modifiers.logo());
+
+    winit::window::Hotkey::new(modifiers, to_virtual_keycode(hotkey.key))
+}
+
+/// Converts a `Menu` from [`iced_native`] to a [`winit`] menu.
+///
+/// [`winit`]: https://github.com/rust-windowing/winit
+/// [`iced_native`]: https://github.com/hecrj/iced/tree/master/native
+pub fn menu<Message>(menu: &Menu<Message>) -> winit::window::Menu {
+    fn menu_i<Message>(
+        converted: &mut winit::window::Menu,
+        starting_id: usize,
+        menu: &Menu<Message>,
+    ) -> usize {
+        let mut id = starting_id;
+
+        for item in menu.iter() {
+            match item {
+                menu::Entry::Item {
+                    content, hotkey, ..
+                } => {
+                    converted.add_item(id, content, hotkey.map(self::hotkey));
+
+                    id += 1;
+                }
+                menu::Entry::Dropdown { content, submenu } => {
+                    let mut converted_submenu = winit::window::Menu::new();
+                    let n_children =
+                        menu_i(&mut converted_submenu, id, submenu);
+
+                    converted.add_dropdown(content, converted_submenu);
+
+                    id += n_children;
+                }
+                menu::Entry::Separator => {
+                    converted.add_separator();
+                }
+            }
+        }
+
+        id - starting_id
+    }
+
+    let mut converted = winit::window::Menu::default();
+    let _ = menu_i(&mut converted, 0, menu);
+
+    converted
+}
+
+/// Given a [`Menu`] and an identifier of a [`menu::Entry`], it returns the
+/// `Message` that should be produced when that entry is activated.
+pub fn menu_message<Message>(menu: &Menu<Message>, id: isize) -> Option<Message>
+where
+    Message: Clone,
+{
+    use std::convert::TryFrom;
+
+    fn find_message<Message>(
+        target: usize,
+        starting_id: usize,
+        menu: &Menu<Message>,
+    ) -> Result<Message, usize>
+    where
+        Message: Clone,
+    {
+        let mut id = starting_id;
+
+        for entry in menu.iter() {
+            match entry {
+                menu::Entry::Item { on_activation, .. } => {
+                    if id == target {
+                        return Ok(on_activation.clone());
+                    }
+
+                    id += 1;
+                }
+                menu::Entry::Dropdown { submenu, .. } => {
+                    match find_message(target, id, submenu) {
+                        Ok(message) => {
+                            return Ok(message);
+                        }
+                        Err(n_children) => {
+                            id += n_children;
+                        }
```

Comment:
While this pattern of "send in IDs -> get out ID and find it" is not my favorite, it seems simple enough to follow the logic. I would prefer there be some sort of `EmplacedEntry` or something that has this ID auto-enumerated by the generating function. But that adds context and state (not to mention winit IDs) out into the Iced layer, which is less desirable than a little bit of ID math.

`/ ramble`